Mplp reworked

kislayabhi opened the pull request:
@abinashpanda , @ankurankan 
I have written the update_message method here. 
Also did some changes in the factor class to meet my needs. (As it is getting reworked so did some hacks there)

kislayabhi committed:
added scalar multiplication and factor summition

kislayabhi committed:
added the mplp class

kislayabhi committed:
added Mplp in __init__.py

ankurankan commented on pgmpy/factors/Factor.py:
```diff
@@ -314,7 +315,26 @@ def product(self, *factors, n_jobs=1):
+            The factor to be added
+        """
+        """
+        :param factor:
```
@kislayabhi Remove this

kislayabhi replied:
@ankurankan : okay.
I needed an add functionality for factors. Any other work around or should i just wait for the reworked Factor?

abinashpanda replied:
@kislayabhi I think by this @ankurankan means to remove this docstring, not this method.

abinashpanda commented on pgmpy/inference/mplp.py:
```diff
@@ -0,0 +1,172 @@
+    def __init__(self, model):
+        super().__init__(model)
+
+        if not isinstance(model, MarkovModel):
```
@kislayabhi check the model before calling the `__init___()` method. Don't initialize the object if there is any error.

abinashpanda commented on pgmpy/inference/mplp.py:
```diff
@@ -0,0 +1,172 @@
+        if not isinstance(model, MarkovModel):
+            raise ValueError('Only MarkovModel is supported')
+
+        # S = { c ∩ c': c, c' ∈ C, c ∩ c' != Null}
```
Don't use unicode characters like ∩. Instead of this you can write its latex expression.

``` latex
S = \{ c \cap c^' : c, c^' \in C, c \cap c^' \neq \emptyset \}
```

abinashpanda commented on pgmpy/inference/mplp.py:
```diff
@@ -0,0 +1,172 @@
+
+        # The generation of all the Intersections of all
+        # the pairwise edges taken one at a time
+        for edge_pair in it.combinations([sorted(i) for i in model.edges()], 2):
```
Do we need to sort the edges. I think we can do without this as well.

abinashpanda commented on pgmpy/inference/mplp.py:
```diff
@@ -0,0 +1,172 @@
+        # Initialize each cluster for variables appearing in each of the edges.
+        # We also need the potentials of the current edges.
+        self.cluster_set = []
+        scope_list = [i.scope() for i in model.get_factors()]
```
Please don't use variables like `i`. You can converted this a

``` python
scope_list = [factor.scope() for factor in model.get_factors()]
```

abinashpanda commented on pgmpy/inference/mplp.py:
```diff
@@ -0,0 +1,172 @@
+
+        Parameters
+        ----------
+        set of variables: tuple
```
Put underscores.

abinashpanda commented on pgmpy/inference/mplp.py:
```diff
@@ -0,0 +1,172 @@
+            self.message_from_cluster = {}
+            for i in self.intersection_sets_for_cluster_c:
+                other_variables = list(set(set_of_variables_for_cluster_c)-i)
+                phi = copy.deepcopy(cluster_potential)
```
@kislayabhi I think instead of using `copy.deepcopy`, we can write a `copy` method in `Factor` class. This method would be returning a new factor using the same variables present in the cluster.

@ankurankan Your suggestions on this?

kislayabhi replied:
Yes. It will make using inplace functions like `maximize()` a lot easier.

abinashpanda commented on pgmpy/inference/mplp.py:
```diff
@@ -0,0 +1,172 @@
+        """
+        updated_results = []
+        # The new updates will take place for the intersection_sets of this cluster.
+        for i in sending_cluster.intersection_sets_for_cluster_c:
```
Please use appropriate variables instead of `i`. In future don't use any variable using i, j, k, l; use proper variable names.

abinashpanda commented on pgmpy/inference/mplp.py:
```diff
@@ -0,0 +1,172 @@
+            cluster_list_except_current = [j for j in self.cluster_set if j != sending_cluster]
+            other_intersects = [j for j in sending_cluster.intersection_sets_for_cluster_c if j != i]
+
+            for cluster in cluster_list_except_current:
```
Why are you creating another list just for the purpose of iterations? Lists are very bad in terms of memory management.

Instead you can use something like this:

``` python
for cluster in self.cluster_set:
    if cluster == sending_cluster:
        pass
```

abinashpanda commented on pgmpy/inference/mplp.py:
```diff
@@ -0,0 +1,172 @@
+                    if i == intersection_set:
+                        message_not_from_cluster.append(cluster.message_from_cluster[intersection_set])
+
+                for other_intersect in other_intersects:
```
Same as above.

abinashpanda commented on pgmpy/inference/mplp.py:
```diff
@@ -0,0 +1,172 @@
+            lambda_1 = copy.deepcopy(sending_cluster.message_from_cluster[i])
+            lambda_1.values = np.zeros(np.prod(lambda_1.cardinality))
+            # lambda_2 = \sum_{}{\lambda_{s'}^{-c}}
+            lambda_2 = copy.deepcopy(lambda_1)
```
@kislayabhi Lets discuss on how to create a method in Factor class for copying. I don't using `copy.deepcopy` method so many times.

kislayabhi replied:
Yup. Copying a Factor would be a one line thing mostly.
We just need to return a new factor with variables, cardinalities and values taken from the given factor.

kislayabhi replied:
Actually I can just use `lambda_2 = Factor(lambda_1.variables, lambda_1.cardinality, lambda_1.values)` if it makes it better memory-wise

abinashpanda commented on pgmpy/inference/mplp.py:
```diff
@@ -0,0 +1,172 @@
+            lambda_c_s = -(1 - 1 / intersection_length) * lambda_1 + (1 / intersection_length) * phi
+            updated_results.append(lambda_c_s)
+
+        for i, j in zip(sending_cluster.intersection_sets_for_cluster_c, range(len(updated_results))):
```
@kislayabhi I don't think its a good solution to modify the input variable passed to this method. It would lead to so many bugs which are not trivial to debug. Please think of some more elegant solution if this can't be removed, else remove this.

kislayabhi committed:
Reduced the use of deep-copy and cleaned the code

kislayabhi commented:
@abinashpanda : This is the same algorithm that I implemented except that Sontag later modified it in his PHD thesis a bit from (the original one in the Fixing-Max Product 2007 paper) that I implemented.

Should I modify for the below one?
Of the set of 3 papers, the Later 2 papers are written by Sontag along with the cpp code so I have an intention to go for that

![image](https://cloud.githubusercontent.com/assets/5661623/8762068/877b3032-2d8b-11e5-8353-56d2b013769c.png)

AFAIK, the modifications are very much trivial

kislayabhi committed:
mplp reworked according to updated equation

kislayabhi committed:
test file added

kislayabhi committed:
added example for Map query

abinashpanda commented on pgmpy/inference/mplp.py:
```diff
@@ -0,0 +1,303 @@
+                self.cluster_set[scope] = self.Cluster(self.intersection_set_variables, factor)
+
+        # L(\delta)
+        self.L = sum([max(self.objective[obj].values) for obj in self.objective])
```
Don't use variable `L`, rename it to `loss` or `delta`, but please don't use `L`.

abinashpanda commented on pgmpy/inference/mplp.py:
```diff
@@ -0,0 +1,303 @@
+        # 1/{\| f \|} max_{x_{f-i}}\left[{\theta_f(x_f) + \sum_{i' in f}{\delta_{i'}^{-f}}(x_i')} \right ]
+
+        # Step. 1) Calculate {\theta_f(x_f) + \sum_{i' in f}{\delta_{i'}^{-f}}(x_i')}
+        objective_cluster = self.objective[tuple(sending_cluster.cluster_variables)]
```
As we have discussed earlier don't use `tuple` instead convert it to a `frozenset`.

abinashpanda commented on pgmpy/tests/test_inference/test_Mplp.py:
```diff
@@ -0,0 +1,100 @@
+    def test_query_single_variable(self):
+        query_result = self.mplp.map_query(1000, 0.0002, 0.0002)
+        # Results from the Sontag code for a mplp run without tightening is:
+        sontag_result = {
```
Don't use the term `sontag_result` over here. Instead rename it to `expected_result` and put the reference of the paper from where you got these values.

abinashpanda commented:
@kislayabhi PR looks good. There are few things that you need to change
- Are you sure the max-character length is 120. I don't think so; if its not the case then make the PEP8 corrections.
- As per our last discussion remove the `Cluster` class and make the changes accordingly.

Rest I don't think we need any major change. Once all these changes are done, this is good to merge.

kislayabhi committed:
few minor fixes

kislayabhi committed:
replaced tuples with frozensets

abinashpanda commented on pgmpy/inference/mplp.py:
```diff
@@ -0,0 +1,302 @@
+                present_variables = list(intersection)
+
+                # Present variables cardinality
+                present_variables_card = [cluster_potential.cardinality[cluster_potential.scope().index(variable)]
```
Don't do such complex operation to just get the cardinality of a variable. Instead you can simply use

``` python
present_variables_card = [cluster_potential.get_cardinality(variable) 
                          for variable in present_variables]
```

abinashpanda commented:
@kislayabhi Everything is good, just few last changes and a another important thing to add is to raise a `NotImplementedException` whenever the user tries to call `query()` method. After that its good to go.

kislayabhi commented:
I will be adding the triplet cluster tightening. Just few hours. Once that is done, we will have the complete implementation of the 1st two papers.

kislayabhi commented:
@abinashpanda , While writing the tests, can we use the new UAI reader to build the model and use it. That way, we can show the users that UAI files are easy to deal with when working with approximate inference.

abinashpanda commented:
@kislayabhi Yeah go for it. It would also be a good integration test.

kislayabhi committed:
added triplet clustering

kislayabhi committed:
added int_gap assertion in test file

kislayabhi committed:
fixed pep8 issues

kislayabhi commented:
@abinashpanda , As mentioned in the paper, we have to break the ties arbitrarily. This causes the result to change and we are left with 2 possible results of which both are correct.

See the test_Mplp file. I hope this is the correct way of handling these issues.

kislayabhi committed:
integrated uai reader in the test

abinashpanda commented on pgmpy/inference/mplp.py:
```diff
@@ -0,0 +1,502 @@
+            * stuck on a local minima.
+
+        Parameters
+        ---------
```
Dashed line should be completely covering the previous line.

abinashpanda commented on pgmpy/inference/mplp.py:
```diff
@@ -0,0 +1,502 @@
+                                   This sets the threshold for the integrality gap below which we say that the solution
+                                   is satisfactory.
+
+        Reference:
```
It should be rather 

```
References
----------
```

abinashpanda commented on pgmpy/inference/mplp.py:
```diff
@@ -0,0 +1,502 @@
+        >>> mplp = Mplp(mm)
+        >>> mplp.find_triangles()
+        """
+        return [clique for clique in nx.find_cliques(self.model) if len(clique) == 3]
```
Use `filter` for this

``` python
return list(filter(lambda x: len(x) == 3, nx.find_cliques(self.model)])
```

This would be much faster as compared to list comprehension with conditions.

abinashpanda commented on pgmpy/inference/mplp.py:
```diff
@@ -0,0 +1,502 @@
+        no_iterations:  integer
+                        Number of maximum iterations that we want MPLP to run.
+        """
+        for niter in range(no_iterations):
```
Rename `niter` to `n_iter`. Trivial changes :P

abinashpanda commented:
@kislayabhi There are a few trivial changes to be made. After that its ready to be merged.

kislayabhi committed:
trivial changes
